include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,519 @@
+//
+// Copyright (C) 2016 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__quadrature_point_data_h
+#define dealii__quadrature_point_data_h
+
+#include <deal.II/base/quadrature.h>
+#include <deal.II/base/config.h>
+#include <deal.II/base/subscriptor.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/tria_accessor.h>
+#include <deal.II/distributed/tria.h>
+#include <deal.II/fe/fe_tools.h>
+#include <deal.II/fe/fe.h>
+#include <deal.II/lac/vector.h>
+
+#include <vector>
+#include <map>
+
+DEAL_II_NAMESPACE_OPEN
+
+/*!@addtogroup Quadrature */
+/*@{*/
+
+
+/**
+ * Base class for storing at each cell @p CELL a vector of quadrature point data @p DATA .
+ *
+ * @author Denis Davydov, Jean-Paul Pelteret, 2016
+ */
+template <typename CELL, typename DATA>
+class QPManagerBase : public Subscriptor
+{
+public:
+  /**
+   * Constructor.
+   */
+  QPManagerBase();
+
+  /**
+   * Destructor.
+   */
+  virtual ~QPManagerBase();
+
+  /**
+   * Initialize quadrature point data for @p cell with @p n_q_points .
+   */
+  void initialize(const CELL &cell,
+                  const unsigned int n_q_points);
+
+  /**
+   * Get reference to the data located at @p cell .
+   */
+  std::vector<DATA> &get_data(const CELL &cell);
+
+  /**
+   * Get constant reference to the data located at @p cell .
+   */
+  const std::vector<DATA> &get_data(const CELL &cell) const;
+
+  /**
+   * Unpack the data stored in @p values_at_qp of @p cell into the underlying
+   * DATA object at each quadrature point.
+   */
+  void unpack_to_cell_data(const CELL &cell,
+                           const FullMatrix<double> &values_at_qp);
+
+  /**
+   * The opposite of the function above. Pack DATA from each quadrature point of the @p cell into @p matrix_data .
+   */
+  void pack_cell_data(const CELL &cell,
+                      FullMatrix<double> &matrix_data);
+
+protected:
+  /**
+   * A map to store a vector of data at the cell.
+   */
+  std::map<CELL, std::vector<DATA> > map;
+
+};
+
+namespace distributed
+{
+
+  /**
+   * A class for storing and managing quadrature points data in with parallel::distributed::Triangulation .
+   *
+   * <h3>Implementation details</h3>
+   *
+   * In addition to the basic functionality provided by the QPManagerBase class,
+   * this class implements transferring of the quadrature point data between
+   * cells if case of adaptive refinement. That also includes automatic shipping of
+   * information between different processors.
+   *
```

Comment:
Make a note that it is assumed that all QP's in a cell have the same construction (specifically, the number of internal data). For this reason, this class may not be sufficiently robust when adopting a level-set approach to describe material behavior.